src/Common/src/System/Buffers/ArrayBufferWriter.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Represents a heap-based, array-backed output sink into which <typeparam name="T"/> data can be written.
+    /// </summary>
+#if USE_ABW_INTERNALLY
+    internal
+#else
+    public
+#endif
+    sealed class ArrayBufferWriter<T> : IBufferWriter<T>
+    {
+        private T[] _buffer;
+        private int _index;
+
+        private const int MinimumBufferSize = 256;
```

Comment:
This isn't actually MinimumBufferSize, at least not how it's currently used.  A better name for the current usage would be DefaultInitialBufferSize, right?

Reply:
Yes, I was thinking the same thing. The only concern is that I use this to define the minimum size GetMemory/GetSpan can return as well.

Reply:
> The only concern is that I use this to define the minimum size GetMemory/GetSpan can return as well.

It isn't the minimum, though. It's just the value that you pretend sizeHint was if it's 0.  (GetSpan(1) is allowed to return a span of Length 1)

Reply:
Fair enough. I meant minimum in the "default" case, but you brought up a useful clarification.